### ISSUE TITLE
macOS client hangs when downloading corrupt compressed blocklists

### ISSUE THREAD

**Issue by GaryElshaw** [transmission/transmission#4702]
### What is the issue?

This bug appears to be back, automated blocklist updating is seemingly not happening, again.

![SCR-20230201-x9v](https://user-images.githubusercontent.com/69029666/216024839-2273b6e6-10c9-4ef0-b230-0503f1c496f2.png)

Edit: An added bonus. When manually updating, 'Processing blocklist...' continues into infinity and the cancel button is greyed out.

![SCR-20230202-4y](https://user-images.githubusercontent.com/69029666/216026207-8babd884-adc8-464f-8a7d-5048c056c7ed.png)

### Which application of Transmission?

macOS app

### Which version of Transmission?

4.0.0-beta.3.dev (724a0f48f0)

**Comment by ckerr** [comment]
Duplicate of https://github.com/transmission/transmission/issues/2272, I think. It's had a `pr welcome` label since June 2022 :smile_cat:

**Comment by GaryElshaw** [comment]
Sadly, no :-( It had been resolved some time ago, but is back. This is a potential blocker. Sucks, i know.

**Comment by ckerr** [comment]
What is the URL of the blocklist that you're using?

**Comment by GaryElshaw** [comment]
What feels like the project default: https://github.com/Naunter/BT_BlockLists

**Comment by ckerr** [comment]
That is not the URL in your screenshot.

I'm trying to reproduce the bug; could you please paste the URL that you're using when you encounter this failure :pray:

**Comment by GaryElshaw** [comment]
It's not the same? https://github.com/Naunter/BT_BlockLists/raw/master/bt_blocklists.gz

**Comment by ckerr** [comment]
Confirmed.

Seems to require `.gz` in order to trigger the bug. 

https://github.com/Naunter/BT_BlockLists/raw/master/bt_blocklists works

https://github.com/Naunter/BT_BlockLists/raw/master/bt_blocklists.gz hangs forever

**Comment by nevack** [comment]
> Naunter/BT_BlockLists@`master`/bt_blocklists.gz (raw)

Ok, I have tried to debug it myself.
This file is invalid tar archive.
@ckerr can you confirm?
Checksum:
`0d25e429de8f1ac5cc51ab3bb99aee326297b545a019e0a58fffc8ade7ae67db  bt_blocklists.gz`
File size:
`680318 bytes`

**Comment by nevack** [comment]
I agree we need to improve this process and handle errors more graceful.

**Comment by ckerr** [comment]
@nevack looks like it's hanging when callling `/usr/bin/tar --list --file $foo` and then waiting on `[tarList waitUntilExit]`. Should we handle that better when fed an invalid archive?

**Comment by nevack** [comment]
> @nevack looks like it's hanging when callling `/usr/bin/tar --list --file $foo` and then waiting on `[tarList waitUntilExit]`. Should we handle that better when fed an invalid archive?

Yep, I see the same hang on waitUntilExit.
I'll take a look what we can do.

I think we should use some lightweight 3rd-party library for working with archives, current implementation (which is my refactor) is a burden :(

**Comment by nevack** [comment]
As per https://stackoverflow.com/questions/2001709/how-to-check-if-a-unix-tar-gz-file-is-a-valid-file-without-uncompressing
we cannot just test archive before listing.
Also, it's really fast check when done by hand in terminal.
I think piping it with NSPipe kills all perfomance.

**Comment by nevack** [comment]
`tar --list --file bt_blocklists.gz | wc -lc` gives me 99457 lines and 4030980 chars.

**Comment by GaryElshaw** [comment]
There are three things. There is the automation/update of the file; how that file is manually updated; but also that the dialog window cannot cancel the process of manually updating.

**Comment by nevack** [comment]
> As per [stackoverflow.com/questions/2001709/how-to-check-if-a-unix-tar-gz-file-is-a-valid-file-without-uncompressing](https://stackoverflow.com/questions/2001709/how-to-check-if-a-unix-tar-gz-file-is-a-valid-file-without-uncompressing) we cannot just test archive before listing. Also, it's really fast check when done by hand in terminal. I think piping it with NSPipe kills all perfomance.

Yup, piping to null fixes this hang.
Let's add check for archive validity.

**Comment by ckerr** [comment]
> Yup, piping to null fixes this hang.

That's what I tried first via the NSTask output property, but wasn't sure how to get the first filename when doing that...

> I think we should use some lightweight 3rd-party library for working with archives, current implementation (which is my refactor) is a burden :(

Agreed! I would love if this was done in core so that it works in non-macOS clients too.

Something like [KArchive](https://api.kde.org/frameworks/karchive/html/index.html) but without the dependency chain...

Maybe https://github.com/libarchive/libarchive ?

**Comment by nevack** [comment]
> That's what I tried first via the NSTask output property, but wasn't sure how to get the first filename when doing that...

I decided to run listing twice - with `stdout > dev/null` and then with `stdout > NSPipe`.
Tested locally with mentioned link to gz.

**Comment by ckerr** [comment]
Yep, WfM too. Thanks for the fast fix @nevack ! :rocket:

**Comment by ckerr** [comment]
@GaryElshaw you're right; this is a blocker. Thanks for reporting this. Also, thanks for persisting when I initially misunderstood the report :+1:
